src/rgw/driver/rados/topic.cc

```diff
@@ -0,0 +1,408 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "topic.h"
+#include "common/errno.h"
+#include "rgw_common.h"
+#include "rgw_metadata.h"
+#include "rgw_metadata_lister.h"
+#include "rgw_notify.h"
+#include "rgw_pubsub.h"
+#include "rgw_rados.h"
+#include "rgw_string.h"
+#include "rgw_tools.h"
+#include "rgw_zone.h"
+#include "svc_mdlog.h"
+#include "svc_sys_obj_cache.h"
+
+namespace rgwrados::topic {
+
+static const std::string oid_prefix = "topic.";
+static constexpr std::string_view buckets_oid_prefix = "buckets.";
+
+static rgw_raw_obj get_topic_obj(const RGWZoneParams& zone,
+                                 std::string_view metadata_key)
+{
+  std::string oid = string_cat_reserve(oid_prefix, metadata_key);
+  return {zone.topics_pool, std::move(oid)};
+}
+
+static rgw_raw_obj get_buckets_obj(const RGWZoneParams& zone,
+                                   std::string_view metadata_key)
+{
+  std::string oid = string_cat_reserve(buckets_oid_prefix, metadata_key);
+  return {zone.topics_pool, std::move(oid)};
+}
+
+
+int read(const DoutPrefixProvider* dpp, optional_yield y,
+         RGWSI_SysObj& sysobj, RGWSI_SysObj_Cache& cache_svc,
+         const RGWZoneParams& zone, const std::string& topic_key,
+         rgw_pubsub_topic& info, RGWChainedCacheImpl<cache_entry>& cache,
+         ceph::real_time* pmtime, RGWObjVersionTracker* pobjv)
+{
+  if (auto e = cache.find(topic_key)) {
+    if (pmtime) {
+      *pmtime = e->mtime;
+    }
+    if (pobjv) {
+      *pobjv = std::move(e->objv);
+    }
+    info = std::move(e->info);
+    return 0;
+  }
+
+  const rgw_raw_obj obj = get_topic_obj(zone, topic_key);
+
+  bufferlist bl;
+  cache_entry entry;
+  rgw_cache_entry_info cache_info;
+  int r = rgw_get_system_obj(&sysobj, obj.pool, obj.oid, bl, &entry.objv,
+                             &entry.mtime, y, dpp, nullptr, &cache_info);
+  if (r < 0) {
+    return r;
+  }
+
+  try {
+    auto p = bl.cbegin();
+    decode(entry.info, p);
+  } catch (const buffer::error&) {
+    return -EIO;
+  }
+
+  cache.put(dpp, &cache_svc, topic_key, &entry, {&cache_info});
```

Comment:
i'm seeing radosgw-admin crashes around this [in teuthology](http://qa-proxy.ceph.com/teuthology/cbodley-2024-02-23_03:33:18-rgw-wip-rgw-meta-topic-distro-default-smithi/7571534/teuthology.log) because read-only radosgw-admin commands like PUBSUB_TOPIC_GET don't create a `cache_svc`
```
2024-02-23T04:45:24.320+0000 7fc9550b4980 20 rados->read ofs=0 len=0
*** Caught signal (Segmentation fault) **
 in thread 7fc9550b4980 thread_name:radosgw-admin
2024-02-23T04:45:24.324+0000 7fc9550b4980 20 rados_obj.operate() r=0 bl.length=226
 ceph version 19.0.0-1678-g23c4a7a4 (23c4a7a41a1bba547fed4ad546ddd748430db41f) squid (dev)
 1: /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7fc95780d520]
 2: __pthread_rwlock_wrlock()
 3: (ObjectCache::chain_cache_entry(DoutPrefixProvider const*, std::initializer_list<rgw_cache_entry_info*>, RGWChainedCache::Entry*)+0x4e) [0x55c47f89197e]
 4: (rgwrados::topic::read(DoutPrefixProvider const*, optional_yield, RGWSI_SysObj&, RGWSI_SysObj_Cache&, RGWZoneParams const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw_pubsub_topic&, RGWChainedCacheImpl<rgwrados::topic::cache_entry>&, std::chrono::time_point<ceph::real_clock, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> > >*, RGWObjVersionTracker*)+0x5aa) [0x55c47f7f1f9a]
 5: (rgw::sal::RadosStore::read_topic_v2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw_pubsub_topic&, RGWObjVersionTracker*, optional_yield, DoutPrefixProvider const*)+0xd4) [0x55c47f722f74]
 6: (RGWPubSub::get_topic(DoutPrefixProvider const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw_pubsub_topic&, optional_yield, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) const+0x6f) [0x55c47f3ffdbf]
 7: main()
 8: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fc9577f4d90]
 9: __libc_start_main()
 10: _start()
```